File: Kotlin/src/com/loggersoft/kotlin/performance/Performer.kt
package com.loggersoft.kotlin.performance

import java.io.File
import java.io.FilenameFilter
import java.io.IOException

class Performer(directory: String?) : FilenameFilter {
    private val dir: File = File(directory ?: ".")
    private val segments: MutableMap<String, MutableList<DcfSegment>> = HashMap()
    private val radius = 5000.0
    private var pos: Position? = null

    override fun accept(dir: File, filename: String): Boolean
            = filename.toLowerCase().endsWith(".dcf")

    fun start() {
        val files = dir.listFiles(this)
        if (files != null) {
            for (f in files) {
                println("Processing ${f.name}...")
                try {
                    DcfFile(f).processFile(::segmentRetrieved, ::invalidLine)
                } catch (e: IOException) {
                    println(e.localizedMessage)
                }
            }
        }
        println("Found ${segments.size} segments;")
    }

    private fun segmentRetrieved(segment: DcfSegment): Boolean {
        pos = pos ?: segment.start
        val position = pos ?: return true
        if (segment.start.getDistance(position) <= radius || segment.end.getDistance(position) <= radius) {
            segments.getOrPut(segment.clustId, { ArrayList() }).add(segment)
        }
        return true
    }

    private fun invalidLine(line: String)
            = println("Invalid DCF line: $line")
}

File: DotNet/Performer.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace CSharp {
    internal class Performer {
        private readonly string _directory;
        private readonly Dictionary<string, List<DcfSegment> > _segments = new Dictionary<string, List<DcfSegment>>();
        private Position _pos;

        public Performer(string directory) {
            _directory = directory ?? ".";
        }

        public void Start() {
            string[] files;
            try {
                files = Directory.GetFiles(_directory, "*.dcf");
            } catch (Exception) {
                return;
            }
            foreach (var file in files) {
                Console.WriteLine("Processing '{0}'...", file.Replace("/", "\\"));
                new DcfFile(file).ProcessFile(SegmentRetreived, InvalidLine);
            }
            Console.WriteLine("Found {0} segments;", _segments.Count);
        }

        private bool SegmentRetreived(DcfSegment segment) {
            if (_pos == null) {
                _pos = segment.Start;
            }
            if (_segments.ContainsKey(segment.ClustId)) {
                _segments[segment.ClustId].Add(segment);
            } else {
                _segments[segment.ClustId] = new List<DcfSegment> { segment };
            }
            return true;
        }

        private static void InvalidLine(string line) {
            Console.Error.WriteLine("Invalid line DCF line: {0}", line);
        }
    }
}
